### PR TITLE
refactor: make `hex_as_bytes` DRY

### PR DESCRIPTION
**Pull request by dhilipsiva**
- Modify `hex_as_bytes` function to return an array of bytes based on input length
- Remove mutable `output` parameter in `hex_as_bytes` function
- Update usage of `hex_as_bytes` function to use generic type and constant size

### CODE REVIEW

**Comment by ibc** [comment]
Please run `cargo fmt`, `cargo clippy` and `cargo test` and commit changes. CI is failing.

**Comment by nazar-pc** [comment]
Just a note: this will bump required Rust compiler version to one that supports const generics (which is relatively old at this point, but still)

**Comment by ds-colligence** [comment]
@ibc I ran all of those 3 commands before pushing the code. `cargo fmt` did not report any bugs. All tests are passing. cliipy does report a few things, but it is not from the code that I touched, it was already there. Should I be fixing those as well?

**Comment by dhilipsiva** [comment]
@ibc I ran all of those 3 commands before pushing the code. cargo fmt did not report any changes. All tests are passing. clippy does report a few things, but it is not from the code that I touched, it was already there. Should I be fixing those as well?

**Comment by ibc** [comment]
Looks like CI or you are running different Rust versions. 

@nazar-pc, how to proceed here? AFAIU mediasoup-rust mandates a minimal Rust version but the user is free to use a major one. Does mediasoup-rust fail to build is running Rust version doesn't satisfy that minimum version declared somewhere in mediasoup files? Should we update that minimum version and match whatever version runs in CI servers?

**Comment by nazar-pc** [comment]
We have `rust-toolchain.toml` in the repo. Assuming Rust toolchain is installed with rustup and Cargo is used, it should be used by default unless overridden. Don't override it, please, use what it chooses out of the box.

P.S. I don't mind bumping version in `rust-toolchain.toml` to much newer release.

**Comment by ibc** [comment]
> We have `rust-toolchain.toml` in the repo. Assuming Rust toolchain is installed with rustup and Cargo is used, it should be used by default unless overridden.

I don't understand this. `rustup` will update Rust to whatever the newest version is, not to the version we say in `rust-toolchain.toml`.

**Comment by ibc** [comment]
@dhilipsiva can you run `rustup --version` in your system? It prints this in my computer:

```
$ rustup --version

rustup 1.26.0 (5af9b9484 2023-04-05)
info: This is the version for the rustup toolchain manager, not the rustc compiler.
info: The currently active `rustc` version is `rustc 1.72.0 (5680fa18f 2023-08-23)`
```

Which matches the Rust version declared in `rust-toolchain.toml` in mediasoup (1.72.0). And when I run all `cargo` commands (fmt, clippy, etc) I do not see any warning, so probably you are using another version of Rust.

**Comment by dhilipsiva** [comment]
@nazar-pc @ibc 

I just realized I have been using rust toolchain installed using brew.

```
rustc --version
rustc 1.78.0 (9b00956e5 2024-04-29) (Homebrew)
```

**Comment by nazar-pc** [comment]
> I don't understand this. `rustup` will update Rust to whatever the newest version is, not to the version we say in `rust-toolchain.toml`.

If `rust-toolchain.toml` is present, that specified toolchain version will be used for the project. Even if default on the system is newer or older. That is the whole point of that file: to have predictable toolchain and set of its components. If you run `rustc --version` in project directory you'll get the same version as in `rust-toolchain.toml`.

**Comment by dhilipsiva** [comment]
@ibc 

I just installed my toolchain using `rustup`.

```
rustup --version
rustup 1.27.1 (54dd3d00f 2024-04-24)
info: This is the version for the rustup toolchain manager, not the rustc compiler.
info: The currently active `rustc` version is `rustc 1.79.0 (129f3b996 2024-06-10)`

rustc --version
rustc 1.79.0 (129f3b996 2024-06-10)
```

* `cargo fmt` reported/changed nothing
* `cargo test` passed
* `cargo clippy` reported parts of code I never touched.

**Comment by dhilipsiva** [comment]
@ibc @nazar-pc I ran all the commands from the root of the repo. Should I have run them from `rust` folder instead?

**Comment by nazar-pc** [comment]
> @ibc @nazar-pc I ran all the commands from the root of the repo. Should I have run them from `rust` folder instead?

If you do that in project directory you should get this:
```
nazar-pc@nazar-pc:/w/g/mediasoup> rustc --version
rustc 1.72.0 (5680fa18f 2023-08-23)
```

Even though global version is different:
```
nazar-pc@nazar-pc:~> rustc --version
rustc 1.77.2 (25ef9e3d8 2024-04-09)
```

Make sure your brew version is not overriding rustup's version.

**Comment by nazar-pc** [comment]
Root works just fine as well

**Comment by dhilipsiva** [comment]
Can confirm that I have uninstalled brew version of rust before running the cargo commands.

```
which rustc
/home/dhilipsiva/.cargo/bin/rustc
which rustup
/home/dhilipsiva/.cargo/bin/rustup
```

**Comment by nazar-pc** [comment]
well, just bump `rust-toolchain.toml` to 1.79.0 then, it'll match your current verison and everything should be fine.

**Comment by ibc** [comment]
> If `rust-toolchain.toml` is present, that specified toolchain version will be used for the project. Even if default on the system is newer or older. That is the whole point of that file: to have predictable toolchain and set of its components. If you run `rustc --version` in project directory you'll get the same version as in `rust-toolchain.toml`.

Imagine I have rustc 1.72.0 in my system and `rust-toolchain.toml` says 1.79.0. Obviously things are gonna fail due to version mismtach.

Imagine I have rustc 1.79.0 in my system and `rust-toolchain.toml` says 1.72.0. What would happen? I don't have rustc 1.72.0. installed in my system.

**Comment by nazar-pc** [comment]
> > If `rust-toolchain.toml` is present, that specified toolchain version will be used for the project. Even if default on the system is newer or older. That is the whole point of that file: to have predictable toolchain and set of its components. If you run `rustc --version` in project directory you'll get the same version as in `rust-toolchain.toml`.
> 
> Imagine I have rustc 1.72.0 in my system and `rust-toolchain.toml` says 1.79.0. Obviously things are gonna fail due to version mismtach.
> 
> Imagine I have rustc 1.79.0 in my system and `rust-toolchain.toml` says 1.72.0. What would happen? I don't have rustc 1.72.0. installed in my system.

It will not fail in either case, It'll install (if not installed already) correct version and will use correct version in the project and nothing else.

**Comment by dhilipsiva** [comment]
> well, just bump `rust-toolchain.toml` to 1.79.0 then, it'll match your current verison and everything should be fine.

Sorry. Should I update the version, commit and push to this branch? Is that what you meant? :sweat_smile:

**Comment by nazar-pc** [comment]
Yes

**Comment by dhilipsiva** [comment]
Update rust version in rust-toolchain.toml to 1.79.0

**Comment by ibc** [comment]
Let's wait for CI.

**Comment by ibc** [comment]
CI failing in `cargo clippy`: https://github.com/versatica/mediasoup/actions/runs/9610107155/job/26507750935?pr=1409

**Comment by ibc** [comment]
> CI failing in `cargo clippy`: https://github.com/versatica/mediasoup/actions/runs/9610107155/job/26507750935?pr=1409

I assume those are due to new version of Rust.

**Comment by dhilipsiva** [comment]
@ibc @nazar-pc Most of the clippy errors were due to planus codegen. I have updated the build script to include `#![allow(clippy::all)` within the generated module to exclude it from clippy. I have done so  using a simple string replace method in the worker build.rs script. I hope this is an acceptable solution. 

Edit: can confirm that all clippy warnings are resolved when tested on my local.

File: worker/build.rs [hunk 1]
```diff
@@ -33,12 +33,13 @@ fn main() {
...
-        format!("{out_dir}/fbs.rs"),
-        planus_codegen::generate_rust(&flatbuffers_declarations)
-            .expect("Failed to generate Rust code from flatbuffers"),
-    )
-    .expect("Failed to write generated Rust flatbuffers into fbs.rs");
+    let planus_content = planus_codegen::generate_rust(&flatbuffers_declarations)
+        .expect("Failed to generate Rust code from flatbuffers");
+
+    let planus_content =
+        planus_content.replace("mod root {", "mod root {\n    #![allow(clippy::all)]");
```
Comment (dhilipsiva) [inline worker/build.rs:40]: @ibc @nazar-pc I hope this is an acceptable solution.

**Comment by nazar-pc** [comment]
Well, the goal is not to get clippy to shut up by suppressing lints/commenting code without any explanation, but to actually address it.

I wanted to fix it, but you have sent PR from an org rather than personal account, so I can't do that. I have pushed commit with fixes here: https://github.com/nazar-pc/mediasoup/tree/dry-DtlsFingerprint-from_fbs-fix-lints

**Comment by dhilipsiva** [comment]
Oh, wow! Using Drop to silence the dead_code error is such a neat trick. :) It didn't cross my mind.

**Comment by nazar-pc** [comment]
Well, another way is to actually use named field (rather than tuple), where name starts with `_`, that way clippy will not complain either. Probably should have done that instead.

**Comment by nazar-pc** [comment]
The key though is to leave an explanation why that is okay/expected

**Comment by ibc** [comment]
So is this ready to merge?

**Comment by nazar-pc** [comment]
I expect one more update as described above

**Comment by dhilipsiva** [comment]
Oh sincere apologies. I misunderstood. I mistakenly assumed that you will merge the changes you made on your fork. Let me do this in a couple of hours.

**Comment by dhilipsiva** [comment]
I have removed ignoring dead_code warnings. I have used named fields as suggested. Please let me know if there is anything I need to change :)

**Comment by dhilipsiva** [comment]
While going through the commit @nazar-pc made on their personal repo, I realized I missed something: 

Silencing clippy lints on files generated by plaun codegen: https://github.com/nazar-pc/mediasoup/commit/7cdf64865afb63f61c68490bf0a20a887e937a42#diff-4bce653f545fe2edb899732221932e371b14c992e8304ab8f6a9677a142a7ec8R4

I did this same exact change and I can confirm that this is what caused #1413 

```
../../../../../../worker/meson.build:284:20: ERROR: Subproject exists but has no meson.build file.
```

I don't think I am smart enough to figure this one out. I deleted the registry folder as mentioned in #1413' s thread - but the same error kept happening.

**Comment by nazar-pc** [comment]
During development files are stored in the project's directory. Delete everything in `worker/subprojects` except wrap files and you should be good to go. `make clean-all` should work in `worker` directory too.

**Comment by ibc** [comment]
Now this is ready to merge, right?

File: rust/src/worker/utils/channel_read_fn.rs [hunk 2]
```diff
@@ -8,9 +8,17 @@ unsafe extern "C" fn free_vec(message: *mut u8, message_len: u32, message_capaci
     Vec::from_raw_parts(message, message_len as usize, message_capacity);
 }
 
-pub(super) struct ChannelReadCallback(
-    Box<dyn (FnMut(UvAsyncT) -> Option<Vec<u8>>) + Send + 'static>,
-);
+pub(super) struct ChannelReadCallback {
+    _callback: Box<dyn (FnMut(UvAsyncT) -> Option<Vec<u8>>) + Send + 'static>,
```
Comment (nazar-pc) [inline rust/src/worker/utils/channel_read_fn.rs:13]: I'd really like for comments to be present at such locations

**Review by nazar-pc — COMMENTED** [review]
Not all comments/suggestions were ported into here still, Inaki

**Comment by ds-colligence** [comment]
I have made all required changes. Please let me know if I have missed any. If there are no more changes and the tests succeeds, we can merge this.

**Review by nazar-pc — APPROVED** [review]
Looks good, thank you for contribution and patience!

**Comment by ibc** [comment]
Merging, thanks a lot.

**Comment by ibc** [comment]
I've updated Rust CHANGELOG for next release: https://github.com/versatica/mediasoup/commit/a80630aa107204fee677e4cf3b11f0f7b22fb064

**Pull request merged**